src/lib/profiles/security/WeaveCert.cpp

```diff
@@ -894,6 +901,509 @@ NL_DLL_EXPORT uint32_t SecondsSinceEpochToPackedCertTime(uint32_t secondsSinceEp
+                                            kKeyUsageFlag_KeyEncipherment;
+    certData->PubKeyAlgoOID               = kOID_PubKeyAlgo_ECPublicKey;
+    certData->SigAlgoOID                  = kOID_SigAlgo_ECDSAWithSHA256;
+    certData->KeyPurposeFlags             = kKeyPurposeFlag_ServerAuth |
+                                            kKeyPurposeFlag_ClientAuth;
+    certData->NotBeforeDate               = WEAVE_CONFIG_DEFAULT_CERT_VALID_DATE_NOT_BEFORE;
+    certData->NotAfterDate                = WEAVE_CONFIG_DEFAULT_CERT_VALID_DATE_NOT_AFTER;
+    certData->PublicKey.EC.ECPoint        = devicePubKey.ECPoint;
+    certData->PublicKey.EC.ECPointLen     = devicePubKey.ECPointLen;
+    certData->EncodedCert                 = cert;
+    certData->EncodedCertLen              = certBufSize;
+
+    CertCount++;
+
+exit:
+    return err;
+}
+
+/**
+ * @brief
+ *   Generate Weave operational device certificate.
+ *
+ * @details
+ *   This function generates Weave self-signed operational certificate encoded in the Weave TLV
+ *   format.
+ *
+ * @param  deviceId                Weave operational device Id.
+ * @param  devicePubKey            Weave operational device public key.
+ * @param  cert                    A pointer to a buffer where generated certificate to be written.
+ * @param  certBufSize             The length in bytes of the provided certificate buffer.
+ * @param  certLen                 The length in bytes of the generated certificate.
+ * @param  genCertSignature        A pointer to a function that generates ECDSA signature on the given
+ *                                 certificate hash using operational device private key.
+ *
+ * @retval #WEAVE_NO_ERROR         If Weave certificate was successfully generated.
+ */
+NL_DLL_EXPORT WEAVE_ERROR GenerateOperationalDeviceCert(uint64_t deviceId, EncodedECPublicKey& devicePubKey,
+                                                        uint8_t *cert, uint16_t certBufSize, uint16_t& certLen,
+                                                        WeaveCertificateSet::GenerateECDSASignatureFunct genCertSignature)
+{
+    enum
+    {
+        kMaxCerts             = 1,      // In this function only one operation device certificate data is loaded.
+        kCertDecodeBufferSize = 1024,   // Maximum ASN1 encoded size of the operational device certificate.
+        kCertKeyIdSize        = 8,      // Size of the certificate key Id value.
+    };
+
+    WEAVE_ERROR err;
+    WeaveCertificateSet certSet;
+    WeaveCertificateData *certData;
+    uint8_t certKeyId[kCertKeyIdSize];
+    bool certSetInitialized = false;
+
+    // Initialize the certificate set.
+    err = certSet.Init(kMaxCerts, kCertDecodeBufferSize, Platform::Security::MemoryAlloc, Platform::Security::MemoryFree);
+    SuccessOrExit(err);
+    certSetInitialized = true;
+
+    // Generate random certificate key Id.
+    err = nl::Weave::Platform::Security::GetSecureRandomData(certKeyId, sizeof(certKeyId));
```

Comment:
This is wrong.  Key ids are hashes of the public key that follow the rules in [RFC5280, section 4.2.1.2](https://tools.ietf.org/html/rfc5280#section-4.2.1.2), clause 2.  You can see the equivalent code for OpenSSL in AddSubjectKeyId() in CertUtils.cpp.

Reply:
Updated implementation